pkg/reconciler/tenancy/initialization/apibinder_initializer_reconcile.go

```diff
@@ -0,0 +1,273 @@
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+	"github.com/kcp-dev/kcp/pkg/logging"
+)
+
+func (b *APIBinder) reconcile(ctx context.Context, clusterWorkspace *tenancyv1alpha1.ClusterWorkspace) error {
+	logger := klog.FromContext(ctx)
+
+	var errors []error
+	clusterName := logicalcluster.From(clusterWorkspace).Join(clusterWorkspace.Name)
+	logger.V(2).Info("initializing APIBindings for workspace")
+
+	// Start with the ClusterWorkspaceType specified by the ClusterWorkspace
+	leafCWT, err := b.getClusterWorkspaceType(logicalcluster.New(clusterWorkspace.Spec.Type.Path), string(clusterWorkspace.Spec.Type.Name))
+	if err != nil {
+		logger.Error(
+			err,
+			"error getting ClusterWorkspaceType",
+			"clusterWorkspaceType.path", clusterWorkspace.Spec.Type.Path,
+			"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+		)
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			tenancyv1alpha1.WorkspaceInitialized,
+			tenancyv1alpha1.WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error getting ClusterWorkspaceType %s|%s: %v",
+			clusterWorkspace.Spec.Type.Path, clusterWorkspace.Spec.Type.Name,
+			err,
+		)
+
+		return nil
+	}
+
+	// Get all the transitive ClusterWorkspaceTypes
+	cwts, err := b.transitiveTypeResolver.Resolve(leafCWT)
+	if err != nil {
+		logger.Error(
+			err,
+			"error resolving transitive types",
+			"clusterWorkspaceType.path", clusterWorkspace.Spec.Type.Path,
+			"clusterWorkspaceType.name", clusterWorkspace.Spec.Type.Name,
+		)
+
+		conditions.MarkFalse(
+			clusterWorkspace,
+			tenancyv1alpha1.WorkspaceInitialized,
+			tenancyv1alpha1.WorkspaceInitializedClusterWorkspaceTypeInvalid,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error resolving transitive set of cluster workspace types: %v",
+			err,
+		)
+
+		return nil
+	}
+
+	// Get current bindings
+	bindings, err := b.listAPIBindings(clusterName)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	// This keeps track of any APIBindings for an APIExport that may have conflicts. We don't really expect this to
+	// happen, because we're tracking what we've created, but "just in case"...
+	bindingsWithConflicts := make(map[tenancyv1alpha1.APIExportReference]sets.String)
+
+	// This keeps track of which APIBindings have been created for which APIExports
+	exportToBinding := map[apisv1alpha1.WorkspaceExportReference]*apisv1alpha1.APIBinding{}
+
+	// Check for any naming conflicts
+	for i := range bindings {
+		binding := bindings[i]
+
+		if binding.Spec.Reference.Workspace == nil {
+			continue
+		}
+
+		// Track what we have ("actual")
+		exportToBinding[*binding.Spec.Reference.Workspace] = binding
+
+		// Record any conflicts
+		for _, cond := range binding.Status.Conditions {
+			if cond.Reason == apisv1alpha1.NamingConflictsReason {
+				logger.V(2).Info("found an APIBinding with conflicts", "message", cond.Message)
```

Comment:
nit: this is the sort of place I think having the APIBinding name etc in the logger would be useful

Reply:
will do